corehq/motech/repeaters/models.py

```diff
@@ -940,31 +972,66 @@ class Meta:
 
     payload_generator_classes = (DataSourcePayloadGenerator,)
 
-    def allowed_to_forward(
-        self,
-        payload,  # type DataSourceUpdateLog
-    ):
-        return payload.data_source_id == self.data_source_id
+    def register(self, payload, fire_synchronously=False):
+        """
+        Merge DataSourceUpdates of waiting repeat records.
+
+        To avoid a race condition where a waiting record is sent while
+        its payload is being updated, doc IDs are added to the new
+        payload and waiting payloads are canceled.
+        """
+        doc_ids = set()
+        for waiting_record in self.repeat_records_ready.all():
+            # We can just fetch all repeat records, because they're all
+            # for the same data source.
+            waiting_record.cancel()  # Cancel to avoid race condition
```

Comment:
Is there any issue if the user manually retriggers these cancelled records? If nothing else, it might be a little confusing for the user since they didn't take an action to cancel these records, and they might not even have any attempts when cancelled. Does it make more sense to just delete them entirely?

Reply:
Yeah, deleting them would probably be better.

Reply:
69bbcb4fb1e